tools/ktx/metrics_utils.h

```diff
@@ -0,0 +1,195 @@
+
+        const auto tSwizzleInfo = determineTranscodeSwizzle(texture, report);
+
+        ktx_error_code_e ec = KTX_SUCCESS;
+        // Decode the encoded texture to observe the compression losses
+        ec = ktxTexture2_TranscodeBasis(texture, KTX_TTF_RGBA32, 0);
+        if (ec != KTX_SUCCESS)
+            report.fatal(rc::KTX_FAILURE, "Failed to transcode KTX2 texture to calculate error metrics: {}", ktxErrorString(ec));
+
+        float overallSSIM[4] = {};
+        float overallPSNR = 0;
+
+        auto refIt = referenceImages.begin();
+        for (uint32_t levelIndex = 0; levelIndex < texture->numLevels; ++levelIndex) {
+            const uint32_t imageWidth = std::max(texture->baseWidth >> levelIndex, 1u);
+            const uint32_t imageHeight = std::max(texture->baseHeight >> levelIndex, 1u);
+            const uint32_t imageDepth = std::max(texture->baseDepth >> levelIndex, 1u);
+
+            for (uint32_t layerIndex = 0; layerIndex < texture->numLayers; ++layerIndex) {
+                for (uint32_t faceIndex = 0; faceIndex < texture->numFaces; ++faceIndex) {
+                    for (uint32_t depthSliceIndex = 0; depthSliceIndex < imageDepth; ++depthSliceIndex) {
+                        assert(refIt != referenceImages.end() && "Internal error");
+
+                        ktx_size_t imageOffset;
+                        ktxTexture_GetImageOffset(texture, levelIndex, layerIndex, faceIndex + depthSliceIndex, &imageOffset);
+                        auto* imageData = texture->pData + imageOffset;
+
+                        rgba8image imageView(imageWidth, imageHeight, reinterpret_cast<rgba8color*>(imageData));
+                        imageView.swizzle(tSwizzleInfo.swizzle);
+
+                        auto& ref = *refIt++;
+
+                        basisu::image basisuImage;
+                        basisuImage.resize(imageWidth, imageHeight);
+
+                        for (uint32_t y = 0; y < imageHeight; ++y) {
+                            for (uint32_t x = 0; x < imageWidth; ++x) {
+                                basisuImage(x, y) = basisu::color_rgba(0, 0, 0, 255);
+                                for (uint32_t c = 0; c < referenceNumChannels; ++c)
+                                    basisuImage(x, y)[c] = imageView(x, y)[c];
+                            }
+                        }
+
+                        if (referenceImages.size() != 1)
+                            fmt::print("Level {}{}{}{}:\n",
+                                    levelIndex,
+                                    texture->isArray ? fmt::format(" Layer {}", layerIndex) : "",
+                                    texture->isCubemap ? fmt::format(" Face {}", faceIndex) : "",
+                                    texture->numDimensions == 3 ? fmt::format(" Depth {}", depthSliceIndex) : "");
+
+                        if (opts.compare_ssim) {
+                            const auto ssim = basisu::compute_ssim(ref, basisuImage, false, false);
```

Comment:
The psnr functions only support unorm8. The ssim function is a float32 operation (with a unorm8 wrapper).
So their format support is very limited.

I dont think the math would allow ssim to be calculated per blocks. psnr might be doable (not a 100% on this), but not with this API.

> It is unfortunate we have to copy into the BasisU image class. We'll have to live with it for now.

Yeah, generic (template) library code would be the solution for this, but that is no possible. `std::mdspan` might help with this in the future to some degree, but even that would require support from the library. And even with mdspan a materialized unorm8 image data would be required which is not the case for most format.

Reply:
Thank you. I am sure expanding on metrics and diff will come up in the future.

If ssim is a float32 operation why does that make "format support ... very limited"?

Reply:
What I meant by limited format support is that psnr is unorm8 only and even if ssnr supports float32 a format conversion is required for the input. I would also worry for formats like unorm32 where conversion to float32 is lossy.

Reply:
Thank you for the explanation.